### PR TITLE
#1795 dump lost equipments file

### LINKED ISSUE #1795: Add the possibility to dump a lost equipments file
Add the possibility to dump a lost equipment file from the .jobs. 
A lost equipment is a component that was connected initially in the network and that is disconnect at the end of the simulation.

<lostEquipments xmlns="http://www.rte-france.com/dynawo">
  <lostEquipment id="_BUS____1-BUS____5-1_AC" type="LINE"/>
</lostEquipments>

### PR DESCRIPTION
**Pull request by valleefred**
Closes #1795

### CODE REVIEW

File: dynawo/sources/Modeler/DataInterface/DYNComponentInterface.cpp [hunk 1]
```diff
@@ -101,6 +101,19 @@ ComponentInterface::getStateVariableReference() {
...
+ComponentInterface::backupComponentVar(const std::string& varName) {
+  int index = getComponentVarIndex(varName);
+  if (index > 0)
+    stateVariables_[index].setBackupValue();
+}
+
+bool
+ComponentInterface::hasComponentVarChanged(const std::string& varName, const double fromValue) const {
+  int index = getComponentVarIndex(varName);
+  return (index > 0) && stateVariables_[index].hasValueChanged(fromValue);
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/DataInterface/DYNComponentInterface.cpp:114]: Why is index=0 excluded?

File: dynawo/sources/Modeler/DataInterface/DYNComponentInterface.cpp [hunk 2]
```diff
@@ -111,6 +124,30 @@ ComponentInterface::getType() const {
   return type_;
 }
 
+const std::string&
+ComponentInterface::getTypeAsString() const {
+  static const std::string stringTypes[] = {
+      "UNKNOWN",
+      "BUS",
+      "CALCULATED_BUS",
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/DataInterface/DYNComponentInterface.cpp:132]: CALCULATED_BUS should be represented by opened switches

File: dynawo/sources/Modeler/DataInterface/DYNComponentInterface.cpp [hunk 3]
```diff
@@ -111,6 +124,30 @@ ComponentInterface::getType() const {
...
+      "SWITCH",
+      "LOAD",
+      "LINE",
+      "GENERATOR",
+      "SHUNT",
+      "DANGLING_LINE",
+      "TWO_WTFO",
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/DataInterface/DYNComponentInterface.cpp:139]: please use TWO_WINDINGS_TRANSFORMER and THREE_WINDINGS_TRANSFORMER

File: dynawo/sources/Modeler/DataInterface/DYNComponentInterface.cpp [hunk 4]
```diff
@@ -111,6 +124,30 @@ ComponentInterface::getType() const {
...
+      "THREE_WTFO",
+      "SVC",
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/DataInterface/DYNComponentInterface.cpp:141]: STATIC_VAR_COMPENSATOR

File: dynawo/sources/Modeler/DataInterface/DYNComponentInterface.cpp [hunk 5]
```diff
@@ -111,6 +124,30 @@ ComponentInterface::getType() const {
...
+      "VSC_CONVERTER",
+      "LCC_CONVERTER",
+      "HVDC_LINE"
+    };
+  ComponentType_t type = type_;
+  if (type < UNKNOWN || type > HVDC_LINE)
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/DataInterface/DYNComponentInterface.cpp:147]: Please add an additional type to the enum (named MAX_TYPE?) to replace this HVDC_LINE that could easily change position in the enum and introduce silent bugs

File: dynawo/sources/Modeler/DataInterface/DYNStateVariable.cpp [hunk 6]
```diff
@@ -203,6 +205,18 @@ StateVariable::setValue(const double& value) {
...
+  rawValue_ = value;
+}
+
+void
+StateVariable::setBackupValue() {
+  backupValue_ = rawValue_;
+}
+
+bool
+StateVariable::hasValueChanged(const double fromValue) const {
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/DataInterface/DYNStateVariable.cpp:218]: Wouldn't it be easier to remove the argument and compare to the backup value?

File: dynawo/sources/Simulation/DYNSimulation.cpp [hunk 7]
```diff
@@ -420,6 +428,18 @@ Simulation::configureFinalStateOutputs() {
...
 
+void
+Simulation::configureLostEquipmentsOutputs() {
+  // Lost equipments settings
+  if (jobEntry_->getOutputsEntry()->getLostEquipmentsEntry()
+      && jobEntry_->getOutputsEntry()->getLostEquipmentsEntry()->getDumpLostEquipments()) {
+    lostEquipmentsCollection_ = LostEquipmentsCollectionFactory::newInstance();
+
+    setLostEquipmentsExportMode(Simulation::EXPORT_LOSTEQUIPMENTS_XML);
+    setLostEquipmentsOutputFile(createAbsolutePath("lostEquipments.xml", outputsDirectory_));
```
Comment (rosiereflo) [inline dynawo/sources/Simulation/DYNSimulation.cpp:439]: Please put this in a  folder similarly to others outputs

File: dynawo/sources/Simulation/DYNSimulation.cpp [hunk 8]
```diff
@@ -805,9 +825,15 @@ Simulation::simulate() {
 
   bool criteriaChecked = true;
   try {
-    if (data_ && (exportIIDM_ || activateCriteria_)) {  // no need to update state variable if the IIDM final state is not exported (same for criteria check)
+    // update state variable only if the IIDM final state is exported, or criteria is checked, or lost equipments are exported
+    if (data_ && (exportIIDM_ || activateCriteria_ || (exportLostEquipmentsMode_ != EXPORT_LOSTEQUIPMENTS_NONE))) {
```
Comment (rosiereflo) [inline dynawo/sources/Simulation/DYNSimulation.cpp:829]: please add a method for this comparison used several times "exportLostEquipmentsMode_ != EXPORT_LOSTEQUIPMENTS_NONE"

File: dynawo/sources/Simulation/DYNSimulationLauncher.cpp [hunk 9]
```diff
@@ -120,6 +120,7 @@ void launchSimu(const std::string& jobsFileName) {
       // Needed as otherwise terminate might crash due to missing staticRef variables
       if (err.key() == DYN::KeyError_t::StateVariableNoReference) {
         simulation->activateExportIIDM(false);
+        simulation->setLostEquipmentsExportMode();
```
Comment (rosiereflo) [inline dynawo/sources/Simulation/DYNSimulationLauncher.cpp:123]: I would prefer to have an explicit argument here instead of a default value

**Review by rosiereflo — CHANGES_REQUESTED** [review]
Please also activate the lost equipments in at least 1 nrt and add the associated reference

File: dynawo/sources/API/LEQ/LEQLostEquipmentsCollection.h [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+   * @brief Const iterator over lost equipments
+   *
+   * Const iterator over lost equipments stored in the collection
+   */
+  class LostEquipmentsCollectionConstIterator {
+   public:
+    /**
+     * @brief Constructor
+     *
+     * Constructor based on constraints. Can create an iterator to the
```
[the remark this answers is not in the document]
Comment (valleefred) [reply dynawo/sources/API/LEQ/LEQLostEquipmentsCollection.h:60]: Fix

File: dynawo/sources/Modeler/DataInterface/DYNComponentInterface.cpp (hunk 1 above)
Comment (valleefred) [reply dynawo/sources/Modeler/DataInterface/DYNComponentInterface.cpp:114]: This is a bug. Same in backupComponentVar.
Fixed !

File: dynawo/sources/Modeler/DataInterface/DYNComponentInterface.cpp (hunk 2 above)
Comment (valleefred) [reply dynawo/sources/Modeler/DataInterface/DYNComponentInterface.cpp:132]: I don't understand the remark !
This type already exists, so it has a string representation too.

File: dynawo/sources/Modeler/DataInterface/DYNComponentInterface.cpp (hunk 3 above)
Comment (valleefred) [reply dynawo/sources/Modeler/DataInterface/DYNComponentInterface.cpp:139]: Done

File: dynawo/sources/Modeler/DataInterface/DYNComponentInterface.cpp (hunk 4 above)
Comment (valleefred) [reply dynawo/sources/Modeler/DataInterface/DYNComponentInterface.cpp:141]: Done

File: dynawo/sources/Modeler/DataInterface/DYNComponentInterface.cpp (hunk 2 above)
Comment (rosiereflo) [reply dynawo/sources/Modeler/DataInterface/DYNComponentInterface.cpp:132]: The calculated_bus is a fictive component generated to represent a set of switches and bus bar sections, thus it has no meaning outside of Dynawo

File: dynawo/sources/Modeler/DataInterface/DYNComponentInterface.cpp (hunk 5 above)
Comment (valleefred) [reply dynawo/sources/Modeler/DataInterface/DYNComponentInterface.cpp:147]: Fixed

File: dynawo/sources/Modeler/DataInterface/DYNStateVariable.cpp (hunk 6 above)
Comment (valleefred) [reply dynawo/sources/Modeler/DataInterface/DYNStateVariable.cpp:218]: No ! This function checks if value has changed **from a specific value**.

File: dynawo/sources/Simulation/DYNSimulationLauncher.cpp (hunk 9 above)
Comment (valleefred) [reply dynawo/sources/Simulation/DYNSimulationLauncher.cpp:123]: Done

File: dynawo/sources/Simulation/DYNSimulation.cpp (hunk 8 above)
Comment (valleefred) [reply dynawo/sources/Simulation/DYNSimulation.cpp:829]: Done

File: dynawo/sources/Simulation/DYNSimulation.cpp (hunk 7 above)
Comment (valleefred) [reply dynawo/sources/Simulation/DYNSimulation.cpp:439]: Done: _lostEquipments.xml_ file in _lostEquipments_ folder

**Comment by valleefred** [comment]
> Please also activate the lost equipments in at least 1 nrt and add the associated reference

Done in following nrt:
- IEEE14_BasicTestCases/IEEE14_DisconnectGroup
- IEEE14_BasicTestCases/IEEE14_DisconnectLine
- IEEE14_WithAutomata/IEEE14_CurrentLimitAutomaton

File: dynawo/sources/Modeler/DataInterface/test/TestDataInterface.cpp [hunk 11]
```diff
@@ -0,0 +1,292 @@
...
+  vl.add(bb.build("MyBus2"));
+
+  IIDM::builders::SwitchBuilder sb;
+  vl.add(sb.build("MySwitch"), "MyBus1", "MyBus2");
+
+  IIDM::builders::LoadBuilder lb;
+  lb.p0(50.);
+  lb.q0(40.);
+  vl.add(lb.build("MyLoad"), c1);
+
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/DataInterface/test/TestDataInterface.cpp:163]: Please add one test in which two components are lost equipments

**Review by rosiereflo — CHANGES_REQUESTED** [review]

File: dynawo/sources/Modeler/DataInterface/test/TestDataInterface.cpp (hunk 11 above)
Comment (valleefred) [reply dynawo/sources/Modeler/DataInterface/test/TestDataInterface.cpp:163]: Test added with three lost equipments

**Review by rosiereflo — APPROVED** [review]

**Comment by rosiereflo** [comment]
A code smell is remaining

**Review by gautierbureau — APPROVED** [review]

**Pull request merged**